Makefile

```diff
@@ -0,0 +1,109 @@
+# here, else it contains an empty string. The net effect is to filter out
+# whether this current run of `make` requires a Python virtual environment.
+NEED_VENV := $(or \
+  $(findstring upgrade,$(MAKECMDGOALS)), \
+  $(findstring requirements,$(MAKECMDGOALS)), \
+  $(findstring all,$(MAKECMDGOALS)), \
+  $(findstring quick-check,$(MAKECMDGOALS)), \
+  $(findstring check,$(MAKECMDGOALS)), \
+  $(findstring test,$(MAKECMDGOALS)), \
+  $(findstring dist,$(MAKECMDGOALS)), \
+  $(findstring bdist-wheel,$(MAKECMDGOALS)), \
+  $(findstring sdist,$(MAKECMDGOALS)), \
+  $(findstring docs,$(MAKECMDGOALS)) \
+)
+ifeq ($(NEED_VENV),)
+  # None of the current goals requires a virtual environment.
+else
+  ifeq ($(wildcard .venv),)
+    $(error No Python environment found, use `make setup` first)
+  else
+    PACKAGE_VERSION=$(shell .venv/bin/python -c 'import package; print(package.__version__)')
+  endif
+endif
+
+.PHONY: setup
+setup:
+	if [ -d .venv ]; then \
+	  echo "Python venv already exists, exiting" && exit 1; \
+	fi
+	if [ -z "${PYTHON}" ]; then \
+	  python3.10 -m venv --upgrade-deps .venv; \
+	else \
+	  ${PYTHON} -m venv --upgrade-deps .venv; \
+	fi
+	. .venv/bin/activate && \
+	pip install --upgrade pip && \
+	pip install --editable .[hooks,dev,test,docs]
+	. .venv/bin/activate && \
+	pre-commit install && \
+	pre-commit install --hook-type commit-msg && \
+	pre-commit install --hook-type pre-push
+
+.PHONY: upgrade
+upgrade:
+	. .venv/bin/activate && \
+	pip install --upgrade pip && \
+	pip install --editable .[hooks,dev,test,docs]
+	rm -fr requirements.txt && $(MAKE) requirements
+
+.PHONY: requirements
+requirements: requirements.txt
+requirements.txt:
+	. .venv/bin/activate && \
+	echo "" > requirements.txt && \
+	for p in `pip list --format freeze`; do hashin --verbose $$p; done
+
+.PHONY: all
+all: check test dist docs
+
+.PHONY: quick-check check
+quick-check:
+	. .venv/bin/activate && \
+	pre-commit run pylint --all-files && \
+	pre-commit run mypy --all-files
+check:
+	. .venv/bin/activate && \
+	pre-commit run --all-files
+
+.PHONY: test
+test:
+	. .venv/bin/activate && \
+	pre-commit run pytest --hook-stage push
+
+.PHONY: dist bdist-wheel sdist
+dist: bdist-wheel sdist
+bdist-wheel: .venv/dist/package-$(PACKAGE_VERSION)-py3-none-any.whl
+.venv/dist/package-$(PACKAGE_VERSION)-py3-none-any.whl:
+	. .venv/bin/activate && \
+	python setup.py bdist_wheel --dist-dir .venv/dist/ --bdist-dir .venv/build
+sdist: .venv/dist/package-$(PACKAGE_VERSION).tar.gz
+.venv/dist/package-$(PACKAGE_VERSION).tar.gz:
+	. .venv/bin/activate && \
+	python setup.py sdist --dist-dir .venv/dist/
+
+.PHONY: docs
+docs: docs/_build/html/index.html
+docs/_build/html/index.html:
+	. .venv/bin/activate && \
+	$(MAKE) -C docs/ html
+
+.PHONY: dist-clean clean
+dist-clean:
+	rm -fr .venv/build/* .venv/dist/*
+clean: dist-clean
+	rm -fr .hypothesis .coverage .mypy_cache .pytest_cache
+	rm -fr docs/_build
+
+.PHONY: nuke-caches nuke
+nuke-caches: clean
```

Comment:
`make nuke-caches` still removes `pre-commit` caches for all my projects :-(

Reply:
Yes I noticed the same, but…

https://github.com/jenstroeger/python-package-template/blob/34b8953e24f1a000098e0377fbb276dc8f829bcd/Makefile#L101-L109

Looking at [the code](https://github.com/pre-commit/pre-commit/blob/main/pre_commit/commands/clean.py), I think the [`clean` option ](https://pre-commit.com/#pre-commit-clean) actually cleans out the _entire_ “store” which impacts all repositories that use pre-commit. Sigh.

Also, I think we should remove the `pre-commit uninstall` because it’s managed by the package setup as part of the `hooks` extra:

https://github.com/jenstroeger/python-package-template/blob/34b8953e24f1a000098e0377fbb276dc8f829bcd/setup.py#L60

Hmm, would be nice if pre-commit would be able to manage/clean caches for a particular repo only 🤔

Reply:
Commit https://github.com/jenstroeger/python-package-template/pull/74/commits/c6dd9925855fffa5cfd257970224a74e41c19481.